File: Recursion/FibonacciNumber.py
# Using Recursion
class Solution:
    def fib(self , n:int)->int:
        if(n<2):
            return n
        return self.fib(n-1)+ self.fib(n-2)

# # Using Memoization
# class Solution:
#     def fib(self, n: int) -> int:
#         cache ={}

#         def recr_fib(n:int)->int:
#             if(n in cache):
#                 return cache[n]
#             if(n<2):
#                 return n

#             result = recr_fib(n-1)+recr_fib(n-2)

#             cache[n] = result
#             return result

#         return recr_fib(n)

File: Recursion/FibonacciNumber.cpp
#include<bits/stdc++.h>
using namespace std;

// Using Recursion
class Solution {
public:
    int fib(int n) {
       if(n<2) return n;
       return fib(n-1) + fib(n-2);
    }
};

// Using memoization
// class Solution {
// public:
//     unordered_map<int , int> cache;
//     int fib(int n) {
//        if(n<2) return n;
//        if(cache[n]!=0) return cache[n];
//        int result =  fib(n-1) + fib(n-2);
//        cache[n] = result;
//        return result;
//     }
// };